compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs

```diff
@@ -273,6 +288,17 @@ impl Context {
             dfg.set_value_from_id(*old_result, *new_result);
         }
     }
+
+    fn get_cached<'a>(
+        dfg: &DataFlowGraph,
+        instruction_result_cache: &'a mut InstructionResultCache,
+        instruction: &Instruction,
+        side_effects_enabled_var: ValueId,
+    ) -> Option<&'a Vec<ValueId>> {
+        let predicate =
+            instruction.requires_acir_gen_predicate(dfg).then_some(side_effects_enabled_var);
+        instruction_result_cache.get(instruction).and_then(|map| map.get(&predicate))
```

Comment:
If the instruction is already cached but without predicate, I believe we can re-use it even if the current instruction is under predicate. However this code uses the cache only if they have the same predicate.

Reply:
Yep, the reason why we haven't done this so far is that we'd need to do some refactoring to get around the borrow checker due to the multiple borrows of the outer hashmap.